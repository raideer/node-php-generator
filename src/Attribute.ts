export type AttributeArgument = string | number | boolean | null;
export type AttributeArguments = Record<string, AttributeArgument>;

export class Attribute {
  private name: string;
  private args: AttributeArguments;

  constructor(name: string, args: AttributeArguments = {}) {
    this.name = name;
    this.args = args;
  }

  public getName(): string {
    return this.name;
  }

  public getArguments(): AttributeArguments {
    return this.args;
  }
}
