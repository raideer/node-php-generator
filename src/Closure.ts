import { asFunction } from "./mixins/asFunction";
import { withAttributes } from "./mixins/withAttributes";
import { Parameter } from "./Parameter";

export class Closure extends asFunction(withAttributes(class {})) {
  private uses: Parameter[] = [];

  getUses(): Parameter[] {
    return this.uses;
  }

  setUses(uses: Parameter[]): this {
    this.uses = uses;

    return this;
  }

  addUse(name: string): this {
    this.uses.push((new Parameter()).setName(name));

    return this;
  }
}
