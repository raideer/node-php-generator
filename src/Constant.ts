import { withAttributes } from "@/mixins/withAttributes";
import { withComments } from "@/mixins/withComments";
import { withName } from "@/mixins/withName";
import { withVisibility } from "@/mixins/withVisibility";

export class Constant extends withVisibility(withAttributes(withComments(withName(class {})))) {
  private value: any;
  private final = false;
  private type: string | null = null;

  setValue(value: any): this {
    this.value = value;

    return this;
  }

  getValue(): any {
    return this.value;
  }

  setFinal(final: boolean): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setType(type: string | null): this {
    this.type = type;

    return this;
  }

  getType(): string | null {
    return this.type;
  }
}

