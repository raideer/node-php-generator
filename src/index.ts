export { Attribute } from './Attribute';
export { ClassType } from './ClassType';
export { Closure } from './Closure';
export { Constant } from './Constant';
export { EnumCase } from './EnumCase';
export { EnumType } from './EnumType';
export { GlobalFunction } from './GlobalFunction';
export { InterfaceType } from './InterfaceType';
export { Literal } from './Literal';
export { Method } from './Method';
export { Parameter } from './Parameter';
export { PhpFile } from './PhpFile';
export { PhpNamespace } from './PhpNamespace';
export { Printer } from './Printer';
export { PromotedParameter } from './PromotedParameter';
export { Property } from './Property';
export { PsrPrinter } from './PsrPrinter';
export { TraitType } from './TraitType';
export { TraitUse } from './TraitUse';
