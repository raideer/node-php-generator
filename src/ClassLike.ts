import { withAttributes } from "@/mixins/withAttributes";
import { withComments } from "@/mixins/withComments";
import { PhpNamespace } from "@/PhpNamespace";

export enum ClassVisibility {
  PUBLIC = "public",
  PROTECTED = "protected",
  PRIVATE = "private",
}

export class ClassLike extends withComments(withAttributes(class {})) {
  private namespace: PhpNamespace | null = null;

  private name: string | undefined;

  constructor(name: string, namespace: PhpNamespace | null = null) {
    super();

    this.name = name;
    this.namespace = namespace;
  }

  getName(): string | undefined {
    return this.name;
  }

  setName(name: string | undefined): this {
    this.name = name;

    return this;
  }

  isClass(): boolean {
    return false;
  }

  isTrait(): boolean {
    return false;
  }

  isInterface(): boolean {
    return false;
  }

  isEnum(): boolean {
    return false;
  }
}
