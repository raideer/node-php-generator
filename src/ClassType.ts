import { ClassLike } from "@/ClassLike";
import { PhpNamespace } from "@/PhpNamespace";
import { Constant } from "@/Constant";
import { Method } from "@/Method";
import { Property } from "@/Property";
import { withConstants } from "@/mixins/withConstants";
import { withMethods } from "@/mixins/withMethods";
import { withProperties } from "@/mixins/withProperties";
import { withTraits } from "@/mixins/withTraits";
import { TraitUse } from "@/TraitUse";

export type ClassTypeMember = Constant | Method | Property | TraitUse;

export class ClassType extends withConstants(withMethods(withProperties(withTraits(ClassLike)))) {
  private final = false;
  private abstract = false;
  private extends: string | null = null;
  private readOnly = false;
  private implements: string[] = [];

  isClass(): boolean {
    return true;
  }

  constructor(name: string, namespace: PhpNamespace | null = null) {
    super(name, namespace);
  }

  setFinal(final: boolean = true): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setAbstract(abstract: boolean = true): this {
    this.abstract = abstract;

    return this;
  }

  setReadOnly(readOnly: boolean = true): this {
    this.readOnly = readOnly;

    return this;
  }

  isReadOnly(): boolean {
    return this.readOnly;
  }

  isAbstract(): boolean {
    return this.abstract;
  }

  setExtends(ext: string | null): this {
    this.extends = ext;

    return this;
  }

  getExtends(): string | null {
    return this.extends;
  }

  setImplements(impl: string[]): this {
    this.implements = impl;

    return this;
  }

  getImplements(): string[] {
    return this.implements;
  }

  addMember(member: ClassTypeMember, override: boolean = false): this {
    const name = member.getName() as string;

    if (member instanceof Constant) {
      this.constants[name] = member;
    } else if (member instanceof Method) {
      this.methods[name] = member;
    } else if (member instanceof Property) {
      this.properties[name] = member;
    } else if (member instanceof TraitUse) {
      this.traits[name] = member;
    }

    return this;
  }
}
