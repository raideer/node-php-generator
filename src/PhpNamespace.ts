import { ClassType } from './ClassType';
import { InterfaceType } from './InterfaceType';
import { TraitType } from './TraitType';
import { EnumType } from './EnumType';
import { GlobalFunction } from './GlobalFunction';
import { Helpers } from './Helpers';

export enum NamespaceType {
  Normal = 'normal',
  Function = 'function',
  Constant = 'constant',
}

export class PhpNamespace {
  private name: string;
  private bracketedSyntax: boolean = false;
  private aliases: Record<string, Record<string, string>> = {
    [NamespaceType.Normal]: {},
    [NamespaceType.Function]: {},
    [NamespaceType.Constant]: {},
  };
  private classes: Record<
    string,
    ClassType | InterfaceType | TraitType | EnumType
  > = {};
  private functions: Record<string, GlobalFunction> = {};

  constructor(name: string) {
    if (name !== '' && !Helpers.isNamespaceIdentifier(name)) {
      throw new Error(`Value '${name}' is not valid name.`);
    }
    this.name = name;
  }

  public getName(): string {
    return this.name;
  }

  public setBracketedSyntax(state: boolean = true): this {
    this.bracketedSyntax = state;
    return this;
  }

  public hasBracketedSyntax(): boolean {
    return this.bracketedSyntax;
  }

  public addUse(
    name: string,
    alias: string | null = null,
    of: string = NamespaceType.Normal
  ): this {
    if (
      !Helpers.isNamespaceIdentifier(name, true) ||
      (Helpers.isIdentifier(name) && Helpers.Keywords[name.toLowerCase()])
    ) {
      throw new Error(
        `Value '${name}' is not valid class/function/constant name.`
      );
    } else if (
      alias &&
      (!Helpers.isIdentifier(alias) || Helpers.Keywords[alias.toLowerCase()])
    ) {
      throw new Error(`Value '${alias}' is not valid alias.`);
    }

    name = name.replace(/^\\/, '');
    const aliases = this.aliases[of];
    const used = of === NamespaceType.Normal ? this.classes : this.functions;

    if (alias === null) {
      let base = Helpers.extractShortName(name);
      let counter = 0;
      do {
        alias = base + (counter ? counter : '');
        counter++;
      } while (
        aliases[alias.toLowerCase()] &&
        aliases[alias.toLowerCase()] !== name
      );
    } else {
      if (
        aliases[alias.toLowerCase()] &&
        aliases[alias.toLowerCase()] !== name
      ) {
        throw new Error(
          `Alias '${alias}' used already for '${aliases[alias.toLowerCase()]}', cannot use for '${name}'.`
        );
      } else if (used[alias.toLowerCase()]) {
        throw new Error(
          `Name '${alias}' used already for '${this.name}\\${used[alias.toLowerCase()].getName()}'.`
        );
      }
    }

    aliases[alias] = name;
    return this;
  }

  public removeUse(name: string, of: string = NamespaceType.Normal): void {
    for (const alias in this.aliases[of]) {
      if (this.aliases[of][alias].toLowerCase() === name.toLowerCase()) {
        delete this.aliases[of][alias];
      }
    }
  }

  public addUseFunction(name: string, alias: string | null = null): this {
    return this.addUse(name, alias, NamespaceType.Function);
  }

  public addUseConstant(name: string, alias: string | null = null): this {
    return this.addUse(name, alias, NamespaceType.Constant);
  }

  public getUses(of: string = NamespaceType.Normal): Record<string, string> {
    const aliases = this.aliases[of];

    // Sort the aliases based on the transformed string
    const sortedAliases = Object.entries(aliases).sort(
      ([aliasA, nameA], [aliasB, nameB]) => {
        const transformedA = nameA.replace(/\\/g, ' ');
        const transformedB = nameB.replace(/\\/g, ' ');
        return transformedA.localeCompare(transformedB);
      }
    );

    // Filter the aliases
    const filteredAliases = sortedAliases.filter(([alias, name]) => {
      const fullAlias = (this.name ? this.name + '\\' : '') + alias;
      return (
        fullAlias.localeCompare(name, undefined, { sensitivity: 'base' }) !== 0
      );
    });

    return Object.fromEntries(filteredAliases);
  }

  public resolveName(name: string, of: string = NamespaceType.Normal): string {
    if (Helpers.Keywords[name.toLowerCase()] || name === '') {
      return name;
    } else if (name[0] === '\\') {
      return name.substring(1);
    }

    const aliases = this.aliases[of];
    if (of !== NamespaceType.Normal) {
      return (
        aliases[name.toLowerCase()] ||
        this.resolveName(Helpers.extractNamespace(name) + '\\') +
          Helpers.extractShortName(name)
      );
    }

    const parts = name.split('\\', 2);
    return (
      aliases[parts[0].toLowerCase()] ||
      this.name + (this.name ? '\\' : '') + name
    );
  }

  public simplifyType(type: string, of: string = NamespaceType.Normal): string {
    return type.replace(/[\w\x7f-\xff\\]+/g, (match) =>
      this.simplifyName(match, of)
    );
  }

  public simplifyName(name: string, of: string = NamespaceType.Normal): string {
    if (Helpers.Keywords[name.toLowerCase()] || name === '') {
      return name;
    }

    name = name.replace(/^\\/, '');

    if (of !== NamespaceType.Normal) {
      for (const alias in this.aliases[of]) {
        if (this.aliases[of][alias].toLowerCase() === name.toLowerCase()) {
          return alias;
        }
      }

      return (
        this.simplifyName(Helpers.extractNamespace(name) + '\\') +
        Helpers.extractShortName(name)
      );
    }

    let shortest: string | null = null;
    let relative = name.startsWith(this.name + '\\')
      ? name.substring(this.name.length + 1)
      : null;

    for (const alias in this.aliases[of]) {
      const original = this.aliases[of][alias];
      if (relative && (relative + '\\').startsWith(alias + '\\')) {
        relative = null;
      }

      if ((name + '\\').startsWith(original + '\\')) {
        const short = alias + name.substring(original.length);
        if (!shortest || shortest.length > short.length) {
          shortest = short;
        }
      }
    }

    if (shortest && relative && shortest.length < relative.length) {
      return shortest;
    }

    return relative ?? shortest ?? (this.name ? '\\' : '') + name;
  }

  public add(
    classType: ClassType | InterfaceType | TraitType | EnumType
  ): this {
    const name = classType.getName();
    if (!name) {
      throw new Error('Class does not have a name.');
    }

    const lower = name.toLowerCase();
    if (this.classes[lower] && this.classes[lower] !== classType) {
      throw new Error(`Cannot add '${name}', because it already exists.`);
    } else if (this.aliases[NamespaceType.Normal][lower]) {
      throw new Error(
        `Name '${name}' used already as alias for ${this.aliases[NamespaceType.Normal][lower]}.`
      );
    }

    this.classes[lower] = classType;
    return this;
  }

  public addClass(name: string): ClassType {
    const classType = new ClassType(name, this);
    this.add(classType);
    return classType;
  }

  public addInterface(name: string): InterfaceType {
    const interfaceType = new InterfaceType(name, this);
    this.add(interfaceType);
    return interfaceType;
  }

  public addTrait(name: string): TraitType {
    const traitType = new TraitType(name, this);
    this.add(traitType);
    return traitType;
  }

  public addEnum(name: string): EnumType {
    const enumType = new EnumType(name, this);
    this.add(enumType);
    return enumType;
  }

  public getClass(
    name: string
  ): ClassType | InterfaceType | TraitType | EnumType {
    const classType = this.classes[name.toLowerCase()];
    if (!classType) {
      throw new Error(`Class '${name}' not found.`);
    }
    return classType;
  }

  public getClasses(): (ClassType | InterfaceType | TraitType | EnumType)[] {
    return Object.values(this.classes);
  }

  public removeClass(name: string): this {
    delete this.classes[name.toLowerCase()];
    return this;
  }

  public addFunction(name: string): GlobalFunction {
    const lower = name.toLowerCase();
    if (this.functions[lower]) {
      throw new Error(`Cannot add '${name}', because it already exists.`);
    } else if (this.aliases[NamespaceType.Function][lower]) {
      throw new Error(
        `Name '${name}' used already as alias for ${this.aliases[NamespaceType.Function][lower]}.`
      );
    }

    const func = new GlobalFunction(name);
    this.functions[lower] = func;
    return func;
  }

  public getFunction(name: string): GlobalFunction {
    const func = this.functions[name.toLowerCase()];
    if (!func) {
      throw new Error(`Function '${name}' not found.`);
    }
    return func;
  }

  public getFunctions(): GlobalFunction[] {
    return Object.values(this.functions);
  }

  public removeFunction(name: string): this {
    delete this.functions[name.toLowerCase()];
    return this;
  }
}
