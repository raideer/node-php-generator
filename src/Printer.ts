import { GlobalFunction } from './GlobalFunction';
import { Closure } from './Closure';
import { Method } from './Method';
import { InterfaceType } from './InterfaceType';
import { ClassType } from './ClassType';
import { TraitType } from './TraitType';
import { EnumType } from './EnumType';
import { PhpNamespace } from './PhpNamespace';
import { PhpFile } from './PhpFile';
import { Dumper } from './Dumper';
import { Helpers } from './Helpers';
import { Attribute } from './Attribute';
import { PromotedParameter } from './PromotedParameter';
import { Constant } from './Constant';
import { Property } from './Property';

export class Printer {
  public wrapLength: number = 120;
  public indentation: string = '    ';
  public linesBetweenProperties: number = 0;
  public linesBetweenMethods: number = 2;
  public linesBetweenUseTypes: number = 0;
  public returnTypeColon: string = ': ';
  public bracesOnNextLine: boolean = true;
  public singleParameterOnOneLine: boolean = false;
  public omitEmptyNamespaces: boolean = true;
  protected namespace: PhpNamespace | null = null;
  protected dumper: Dumper;
  private resolveTypes: boolean = true;

  constructor() {
    this.dumper = new Dumper();
  }

  public printFunction(
    func: GlobalFunction,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const line =
      'function ' + (func.getReturnReference() ? '&' : '') + func.getName();
    const returnType = this.printReturnType(func);
    const params = this.printParameters(
      func,
      line.length + returnType.length + 2
    );
    const body = Helpers.simplifyTaggedNames(func.getBody(), this.namespace);
    const braceOnNextLine = this.isBraceOnNextLine(
      params.includes('\n'),
      !!returnType
    );

    return (
      this.printDocComment(func) +
      this.printAttributes(func.getAttributes()) +
      line +
      params +
      returnType +
      (braceOnNextLine ? '\n' : ' ') +
      '{\n' +
      this.indent(body) +
      '}\n'
    );
  }

  public printClosure(
    closure: Closure,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const uses = closure
      .getUses()
      .map((param) => (param.isReference() ? '&' : '') + '$' + param.getName());
    const useStr =
      uses.length > 1 && uses.join(', ').length > this.wrapLength
        ? '\n' + this.indentation + uses.join(',\n' + this.indentation) + ',\n'
        : uses.join(', ');
    const body = Helpers.simplifyTaggedNames(closure.getBody(), this.namespace);

    return (
      this.printAttributes(closure.getAttributes(), true) +
      'function ' +
      (closure.getReturnReference() ? '&' : '') +
      this.printParameters(closure) +
      (uses.length ? ' use (' + useStr + ')' : '') +
      this.printReturnType(closure) +
      ' {\n' +
      this.indent(body) +
      '}'
    );
  }

  public printMethod(
    method: Method,
    namespace: PhpNamespace | null = null,
    isInterface: boolean = false
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    method.validate();
    const line =
      (method.isAbstract() && !isInterface ? 'abstract ' : '') +
      (method.isFinal() ? 'final ' : '') +
      (method.getVisibility() ? method.getVisibility() + ' ' : '') +
      (method.isStatic() ? 'static ' : '') +
      'function ' +
      (method.getReturnReference() ? '&' : '') +
      method.getName();
    const returnType = this.printReturnType(method);
    const params = this.printParameters(
      method,
      line.length + returnType.length + this.indentation.length + 2
    );
    let body = Helpers.simplifyTaggedNames(method.getBody(), this.namespace);
    body = Helpers.normalize(body).trimStart();
    body = body.trimEnd() + '\n';
    const braceOnNextLine = this.isBraceOnNextLine(
      params.includes('\n'),
      !!returnType
    );

    return (
      this.printDocComment(method) +
      this.printAttributes(method.getAttributes()) +
      line +
      params +
      returnType +
      (method.isAbstract() || isInterface
        ? ';\n'
        : (braceOnNextLine ? '\n' : ' ') + '{\n' + this.indent(body) + '}\n')
    );
  }

  public printClass(
    classType: ClassType | InterfaceType | TraitType | EnumType,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    classType.validate();
    const resolver = this.namespace
      ? (s: string) => this.namespace!.simplifyType(s)
      : (s: string) => s;

    const traits =
      classType instanceof ClassType ||
      classType instanceof TraitType ||
      classType instanceof EnumType
        ? classType
            .getTraits()
            .map((trait) => {
              const resolutions = Helpers.simplifyTaggedNames(
                trait.getResolutions().join(';\n'),
                this.namespace
              );
              return (
                this.printDocComment(trait) +
                'use ' +
                resolver(trait.getName()) +
                (resolutions
                  ? ' {\n' + this.indent(resolutions) + ';\n}\n'
                  : ';\n')
              );
            })
            .join('')
        : '';

    const cases =
      classType instanceof EnumType
        ? classType
            .getCases()
            .map((enumCase) => {
              return (
                this.printDocComment(enumCase) +
                this.printAttributes(enumCase.getAttributes()) +
                'case ' +
                enumCase.getName() +
                (enumCase.getValue() === null
                  ? ''
                  : ' = ' + this.dump(enumCase.getValue())) +
                ';\n'
              );
            })
            .join('')
        : '';

    const readOnlyClass = classType instanceof ClassType;
    const consts = classType
      .getConstants()
      .map((constant) => this.printConstant(constant))
      .join('');
    const methods = classType
      .getMethods()
      .map((method) => {
        if (readOnlyClass && method.getName() === 'constructor') {
          const clonedMethod = method.clone();

          clonedMethod.getParameters().forEach((param) => {
            if (param instanceof PromotedParameter) {
              param.setReadOnly(false);
            }
          });

          return this.printMethod(
            clonedMethod,
            namespace,
            classType.isInterface()
          );
        }

        return this.printMethod(method, namespace, classType.isInterface());
      })
      .join('');
    const properties =
      classType instanceof ClassType || classType instanceof TraitType
        ? classType
            .getProperties()
            .map((property) => this.printProperty(property, readOnlyClass))
            .join('')
        : '';

    const members = [traits, consts, cases, properties, methods]
      .filter(Boolean)
      .join('\n');

    let classExtends: string[] = [];

    if (classType instanceof ClassType || classType instanceof InterfaceType) {
      if (classType.getExtends()) {
        if (classType instanceof ClassType) {
          classExtends = [classType.getExtends()!];
        } else {
          classExtends = classType.getExtends();
        }
      }
    }

    const line = [
      classType instanceof ClassType && classType.isAbstract()
        ? 'abstract'
        : null,
      classType instanceof ClassType && classType.isFinal() ? 'final' : null,
      classType instanceof ClassType && classType.isReadOnly()
        ? 'readonly'
        : null,
      classType instanceof ClassType ? 'class ' + classType.getName() : null,
      classType instanceof InterfaceType
        ? 'interface ' + classType.getName()
        : null,
      classType instanceof TraitType ? 'trait ' + classType.getName() : null,
      classType instanceof EnumType
        ? 'enum ' +
          classType.getName() +
          (classType.getType()
            ? this.returnTypeColon + classType.getType()
            : '')
        : null,
      classExtends.length
        ? 'extends ' + classExtends.map(resolver).join(', ')
        : null,
      (classType instanceof ClassType || classType instanceof EnumType) &&
      classType.getImplements().length
        ? 'implements ' + classType.getImplements().map(resolver).join(', ')
        : null,
    ]
      .filter(Boolean)
      .join(' ');

    return (
      this.printDocComment(classType) +
      this.printAttributes(classType.getAttributes()) +
      line +
      '\n{\n' +
      this.indent(members) +
      '}\n'
    );
  }

  public printEnum(enumType: EnumType): string {
    return this.printClass(enumType);
  }

  public printTrait(traitType: TraitType): string {
    return this.printClass(traitType);
  }

  public printInterface(interfaceType: InterfaceType): string {
    return this.printClass(interfaceType);
  }

  public printNamespace(namespace: PhpNamespace): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const name = namespace.getName();
    const uses = [
      this.printUses(namespace),
      this.printUses(namespace, PhpNamespace.NameFunction),
      this.printUses(namespace, PhpNamespace.NameConstant),
    ]
      .filter(Boolean)
      .join('\n'.repeat(this.linesBetweenUseTypes));

    const items = [
      ...namespace.getClasses().map((cls) => this.printClass(cls, namespace)),
      ...namespace
        .getFunctions()
        .map((fn) => this.printFunction(fn, namespace)),
    ].join('\n');

    const body = (uses ? uses + '\n' : '') + items;

    return namespace.hasBracketedSyntax()
      ? `namespace${name ? ` ${name}` : ''} {\n${this.indent(body)}}\n`
      : `${name ? `namespace ${name};\n\n` : ''}${body}`;
  }

  public printFile(file: PhpFile): string {
    const namespaces = file
      .getNamespaces()
      .map((ns) => {
        if (
          !this.omitEmptyNamespaces ||
          ns.getClasses().length ||
          ns.getFunctions().length
        ) {
          return this.printNamespace(ns);
        }
        return '';
      })
      .filter(Boolean)
      .join('\n\n');

    return (
      `<?php\n` +
      (file.getComment() ? '\n' + this.printDocComment(file) : '') +
      '\n' +
      (file.hasStrictTypes() ? 'declare(strict_types=1);\n\n' : '') +
      namespaces
    );
  }

  protected printUses(
    namespace: PhpNamespace,
    of: string = PhpNamespace.NameNormal
  ): string {
    const prefix = {
      [PhpNamespace.NameNormal]: '',
      [PhpNamespace.NameFunction]: 'function ',
      [PhpNamespace.NameConstant]: 'const ',
    }[of];
    return Object.entries(namespace.getUses(of))
      .map(([alias, original]) => {
        return Helpers.extractShortName(original) === alias
          ? `use ${prefix}${original};\n`
          : `use ${prefix}${original} as ${alias};\n`;
      })
      .join('');
  }

  protected printParameters(
    func: GlobalFunction | Method | Closure,
    column: number = 0
  ): string {
    const special = func
      .getParameters()
      .some(
        (param) =>
          param instanceof PromotedParameter ||
          param.getAttributes().length ||
          param.getComment()
      );

    if (
      !special ||
      (this.singleParameterOnOneLine && func.getParameters().length === 1)
    ) {
      const line = this.formatParameters(func, false);
      if (!line.includes('\n') && line.length + column <= this.wrapLength) {
        return line;
      }
    }
    return this.formatParameters(func, true);
  }

  private formatParameters(
    func: GlobalFunction | Method | Closure,
    multiline: boolean
  ): string {
    const params = func.getParameters();

    const res = params
      .map((param) => {
        const variadic =
          func.isVariadic() && param === params[params.length - 1];
        const attrs = this.printAttributes(param.getAttributes(), true);

        return (
          this.printDocComment(param) +
          (attrs ? (multiline ? attrs.slice(0, -1) + '\n' : attrs) : '') +
          (param instanceof PromotedParameter
            ? (param.getVisibility() || 'public') +
              (param.isReadOnly() && param.getType() ? ' readonly' : '') +
              ''
            : '') +
          this.printType(param.getType(), param.isNullable()) +
          ' ' +
          (param.isReference() ? '&' : '') +
          (variadic ? '...' : '') +
          '$' +
          param.getName() +
          (param.getDefaultValue() && !variadic
            ? ' = ' + this.dump(param.getDefaultValue())
            : '') +
          (multiline ? ',\n' : ', ')
        );
      })
      .join('');

    return multiline
      ? '(\n' + this.indent(res) + ')'
      : '(' + res.slice(0, -2) + ')';
  }

  private printConstant(constant: Constant): string {
    const def =
      (constant.isFinal() ? 'final ' : '') +
      (constant.getVisibility() ? constant.getVisibility() + ' ' : '') +
      'const ' +
      this.printType(constant.getType(), false) +
      ' ' +
      constant.getName() +
      ' = ';

    return (
      this.printDocComment(constant) +
      this.printAttributes(constant.getAttributes()) +
      def +
      this.dump(constant.getValue()) +
      ';\n'
    );
  }

  private printProperty(
    property: Property,
    readOnlyClass: boolean = false
  ): string {
    property.validate();
    const type = property.getType();
    const def =
      (property.getVisibility() || 'public') +
      (property.isStatic() ? ' static' : '') +
      (!readOnlyClass && property.isReadOnly() && type ? ' readonly' : '') +
      ' ' +
      this.printType(type, property.isNullable()) +
      ' ' +
      '$' +
      property.getName();

    return (
      this.printDocComment(property) +
      this.printAttributes(property.getAttributes()) +
      def +
      (property.getValue() === null && !property.isInitialized()
        ? ''
        : ' = ' + this.dump(property.getValue())) +
      ';\n'
    );
  }

  protected printType(type: string | undefined, nullable: boolean): string {
    if (!type) return '';
    if (this.namespace) {
      type = this.namespace.simplifyType(type);
    }
    return nullable ? '?' + type : type;
  }

  protected printDocComment(commentable: {
    getComment(): string | undefined;
  }): string {
    const multiLine =
      commentable instanceof GlobalFunction ||
      commentable instanceof Method ||
      commentable instanceof ClassType ||
      commentable instanceof PhpFile;
    return Helpers.formatDocComment(commentable.getComment() || '', multiLine);
  }

  protected printReturnType(func: GlobalFunction | Method | Closure): string {
    const type = this.printType(func.getReturnType(), func.isReturnNullable());
    return type ? this.returnTypeColon + type : '';
  }

  protected printAttributes(
    attrs: Attribute[],
    inline: boolean = false
  ): string {
    if (!attrs.length) return '';
    this.dumper.indentation = this.indentation;
    const items = attrs.map((attr) => {
      const args = this.dumper.format('...?:', attr.getArguments());
      return this.printType(attr.getName(), false) + (args ? `(${args})` : '');
    });
    return inline ? `#[${items.join(', ')}] ` : `#[${items.join(']\n#[')}]\n`;
  }

  protected indent(s: string): string {
    s = s.replace(/\t/g, this.indentation);
    return s.replace(/(?:^|[\r\n]+)(?=[^\r\n])/g, '$&' + this.indentation);
  }

  protected dump(varToDump: any, column: number = 0): string {
    this.dumper.indentation = this.indentation;
    this.dumper.wrapLength = this.wrapLength;
    let s = this.dumper.dump(varToDump, column);
    s = Helpers.simplifyTaggedNames(s, this.namespace);
    return s;
  }

  protected isBraceOnNextLine(
    multiLine: boolean,
    hasReturnType: boolean
  ): boolean {
    return this.bracesOnNextLine && (!multiLine || hasReturnType);
  }
}
