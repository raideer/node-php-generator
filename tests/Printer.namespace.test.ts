import { Printer } from '../src/Printer';
import { PhpNamespace } from '../src/PhpNamespace';
import { ClassType } from '../src/ClassType';

test('prints empty namespace', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App');

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints namespace with use statements', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App\\Entity');
  namespace.addUse('Doctrine\\ORM\\Mapping', 'ORM');
  namespace.addUse('App\\Traits\\TimestampableTrait');
  namespace.addUse('App\\Interfaces\\UserInterface');

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints namespace with aliased use statements', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App');
  namespace.addUse('Foo\\Bar\\Baz', 'CustomBaz');
  namespace.addUse('Foo\\Bar\\Qux', 'CustomQux');

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints namespace with class', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App');
  const classType = new ClassType('User');
  classType.setExtends('BaseEntity');
  namespace.add(classType);

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints namespace with multiple classes', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App\\Entity');

  const class1 = new ClassType('User');
  const class2 = new ClassType('Post');
  namespace.add(class1);
  namespace.add(class2);

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints namespace with bracketed syntax', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App');
  namespace.setBracketedSyntax(true);

  const classType = new ClassType('User');
  namespace.add(classType);

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});

test('prints complex namespace', () => {
  const printer = new Printer();
  const namespace = new PhpNamespace('App\\Entity');

  // Add use statements
  namespace.addUse('Doctrine\\ORM\\Mapping', 'ORM');
  namespace.addUse('App\\BaseEntity');
  namespace.addUse('App\\Traits\\TimestampableTrait');
  namespace.addUse('App\\Interfaces\\EntityInterface');
  namespace.addUse('App\\Repository\\UserRepository');

  // Add class
  const classType = new ClassType('User');
  classType.setExtends('App\\BaseEntity');
  classType.addTrait('App\\Traits\\TimestampableTrait');
  classType.setImplements(['App\\Interfaces\\EntityInterface']);
  namespace.add(classType);

  // Use bracketed syntax
  namespace.setBracketedSyntax(true);

  const output = printer.printNamespace(namespace);

  expect(output).toMatchSnapshot();
});
