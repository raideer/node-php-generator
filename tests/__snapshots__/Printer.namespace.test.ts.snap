// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints complex namespace 1`] = `
"namespace App\\Entity {
    use App\\BaseEntity;
    use App\\Interfaces\\EntityInterface;
    use App\\Repository\\UserRepository;
    use App\\Traits\\TimestampableTrait;
    use Doctrine\\ORM\\Mapping as ORM;

    class User extends BaseEntity implements EntityInterface
    {
        use TimestampableTrait;
    }
}
"
`;

exports[`prints empty namespace 1`] = `
"namespace App;

"
`;

exports[`prints namespace with aliased use statements 1`] = `
"namespace App;

use Foo\\Bar\\Baz as CustomBaz;
use Foo\\Bar\\Qux as CustomQux;

"
`;

exports[`prints namespace with bracketed syntax 1`] = `
"namespace App {
    class User
    {
    }
}
"
`;

exports[`prints namespace with class 1`] = `
"namespace App;

class User extends \\BaseEntity
{
}
"
`;

exports[`prints namespace with multiple classes 1`] = `
"namespace App\\Entity;

class User
{
}

class Post
{
}
"
`;

exports[`prints namespace with use statements 1`] = `
"namespace App\\Entity;

use App\\Interfaces\\UserInterface;
use App\\Traits\\TimestampableTrait;
use Doctrine\\ORM\\Mapping as ORM;

"
`;
